crypto/threads_win.c

```diff
@@ -37,20 +42,347 @@ typedef struct {
+            return;
+        }
+    }
+}
+
+static struct rcu_qp *update_qp(CRYPTO_RCU_LOCK lock)
+{
+    uint64_t new_id;
+    uint32_t current_idx;
+    uint32_t tmp;
+
+    EnterCriticalSection(&lock->alloc_lock);
+    /*
+     * we need at least one qp to be available with one
+     * left over, so that readers can start working on
+     * one that isn't yet being waited on
+     */
+    while (lock->group_count - lock->writers_alloced < 2)
+        SleepConditionVariableCS(&lock->alloc_signal, &lock->alloc_lock, INFINITE);
+
+    current_idx = lock->current_alloc_idx;
+    /* Allocate the qp */
+    lock->writers_alloced++;
+
+    /* increment the allocation index */
+    lock->current_alloc_idx =
+        (lock->current_alloc_idx + 1) % lock->group_count;
+
+    /* get and insert a new id */
+    new_id = lock->id_ctr;
+    lock->id_ctr++;
+
+    new_id = VAL_ID(new_id);
+    InterlockedAnd64(&lock->qp_group[current_idx].users, ID_MASK);
+    InterlockedAdd64(&lock->qp_group[current_idx].users, new_id);
+
+    /* update the reader index to be the prior qp */
+    tmp = lock->current_alloc_idx;
+    InterlockedExchange(&lock->reader_idx, tmp);
+
+    /* wake up any waiters */
+    WakeAllConditionVariable(&lock->alloc_signal);
+    LeaveCriticalSection(&lock->alloc_lock);
+    return (struct rcu_qp *)&lock->qp_group[current_idx];
+}
+
+static void retire_qp(CRYPTO_RCU_LOCK lock,
+                      struct rcu_qp *qp)
+{
+    EnterCriticalSection(&lock->alloc_lock);
+    lock->writers_alloced--;
+    WakeAllConditionVariable(&lock->alloc_signal);
+    LeaveCriticalSection(&lock->alloc_lock);
+}
+
+
+void ossl_synchronize_rcu(CRYPTO_RCU_LOCK lock)
+{
+    struct rcu_qp *qp;
+    uint64_t count;
+    struct rcu_cb_item *cb_items, *tmpcb;
+
+    /* before we do anything else, lets grab the cb list */
+    cb_items = InterlockedExchangePointer((void * volatile *)&lock->cb_items, NULL);
+
+    qp = update_qp(lock);
+
+    /* wait for the reader count to reach zero */
+    do {
+        count = InterlockedOr64(&qp->users, 0);
+    } while (READER_COUNT(count) != 0);
+
+    /* retire in order */
+    EnterCriticalSection(&lock->prior_lock);
+    while (lock->next_to_retire != ID_VAL(count))
+        SleepConditionVariableCS(&lock->prior_signal, &lock->prior_lock, INFINITE);
+
+    lock->next_to_retire++;
+    WakeAllConditionVariable(&lock->prior_signal);
+    LeaveCriticalSection(&lock->prior_lock);
+
+    retire_qp(lock, qp);
+
+    /* handle any callbacks that we have */
+    while (cb_items != NULL) {
+        tmpcb = cb_items;
+        cb_items = cb_items->next;
+        tmpcb->fn(tmpcb->data);
+        OPENSSL_free(tmpcb);
+    }
+
+    /* and we're done */
+    return;
+
+}
+
+int ossl_rcu_call(CRYPTO_RCU_LOCK lock, rcu_cb_fn cb, void *data)
+{
+	struct rcu_cb_item *new;
+	struct rcu_cb_item *prev;
+
+    new = OPENSSL_zalloc(sizeof(struct rcu_cb_item));
+    if (new != NULL)
+        return 0; 
+    prev = new;
```

Comment:
nit: indent

Reply:
not sure I follow.  The only indent I see there is the return 0 statement, and it appears to be indented properly